controllers/kubeadm_control_plane_controller.go

```diff
@@ -104,13 +107,86 @@ func (r *KubeadmControlPlaneReconciler) Reconcile(req ctrl.Request) (res ctrl.Re
 }
 
 // reconcile handles KubeadmControlPlane reconciliation.
-func (r *KubeadmControlPlaneReconciler) reconcile(_ context.Context, kubeadmControlPlane *clusterv1.KubeadmControlPlane, logger logr.Logger) ctrl.Result {
+func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, kcp *clusterv1.KubeadmControlPlane, logger logr.Logger) ctrl.Result {
 	// If object doesn't have a finalizer, add one.
-	controllerutil.AddFinalizer(kubeadmControlPlane, clusterv1.KubeadmControlPlaneFinalizer)
+	controllerutil.AddFinalizer(kcp, clusterv1.KubeadmControlPlaneFinalizer)
 
+	// Fetch the Cluster.
+	cluster, err := util.GetOwnerCluster(ctx, r.Client, kcp.ObjectMeta)
+	if err != nil {
+		logger.Error(err, "Failed to retrieve owner Cluster from the API Server")
+		return ctrl.Result{Requeue: true}
+	}
+	if cluster == nil {
+		logger.Info("Cluster Controller has not yet set OwnerRef")
+		return reconcile.Result{}
+	}
+	logger = logger.WithValues("cluster", cluster.Name)
+
+	labelSelector := generateKubeadmControlPlaneSelector(cluster.Name)
+	selector, err := metav1.LabelSelectorAsSelector(labelSelector)
+	if err != nil {
+		// Since we are building up the LabelSelector above, this should not fail
+		logger.Error(err, "failed to parse label selector")
+		return ctrl.Result{Requeue: true}
+	}
+	// Copy label selector to its status counterpart in string format.
+	// This is necessary for CRDs including scale subresources.
+	kcp.Status.Selector = selector.String()
+
+	selectorMap, err := metav1.LabelSelectorAsMap(labelSelector)
+	if err != nil {
+		// Since we are building up the LabelSelector above, this should not fail
+		logger.Error(err, "failed to convert label selector to a map")
+		return ctrl.Result{Requeue: true}
+	}
+
+	// Get all Machines linked to this KubeadmControlPlane.
+	allMachines := &clusterv1.MachineList{}
+	err = r.Client.List(
+		context.Background(), allMachines,
```

Comment:
nit: I'd either put all the args on one line, or if it's too long, do one arg per line. Having two here looks strange to me.